core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1424,11 +1424,18 @@ class UnifiedLog(@volatile var logStartOffset: Long,
 
   /**
     * Given a message offset, find its corresponding offset metadata in the log.
-    * If the message offset is out of range, throw an OffsetOutOfRangeException
+    * 1. If the message offset is lesser than the log-start-offset, then throw an OffsetOutOfRangeException
+    * 2. If the message offset is lesser than the local-log-start-offset, then it returns the message-only metadata
+    * 3. If the message offset is greater than the log-end-offset, then it returns the message-only metadata
     */
-  private def convertToOffsetMetadataOrThrow(offset: Long): LogOffsetMetadata = {
+  private[log] def convertToOffsetMetadataOrThrow(offset: Long): LogOffsetMetadata = {
```

Comment:
`LocalLog.read()` also calls `convertToOffsetMetadataOrThrow`.

```
      else if (startOffset > maxOffsetMetadata.messageOffset)
        emptyFetchDataInfo(convertToOffsetMetadataOrThrow(startOffset), includeAbortedTxns)
```
It seems this could lead to infinite recursion. To avoid that, we could change the above code to avoid calling `convertToOffsetMetadataOrThrow` and return a message-only `LogOffsetMetadata` instead to `emptyFetchDataInfo`.

Reply:
This is a potential loop (not sure when it would be triggered), updated the logic to return the message-only metadata.

Reply:
To solve the infinite loop, instead of returning the message-only `LogOffsetMetadata`. Can we retain the same behavior without the loop? Something like below:

```java
def read(startOffset: Long,
           maxLength: Int,
           minOneMessage: Boolean,
           maxOffsetMetadata: LogOffsetMetadata,
           includeAbortedTxns: Boolean): FetchDataInfo = {
    maybeHandleIOException(s"Exception while reading from $topicPartition in dir ${dir.getParent}") {
      trace(s"Reading maximum $maxLength bytes at offset $startOffset from log with " +
        s"total length ${segments.sizeInBytes} bytes")

      val endOffsetMetadata = nextOffsetMetadata
      val endOffset = endOffsetMetadata.messageOffset
      val segmentOpt = segments.floorSegment(startOffset)

      // return error on attempt to read beyond the log end offset
      if (startOffset > endOffset || !segmentOpt.isPresent)
        throw new OffsetOutOfRangeException(s"Received request for offset $startOffset for partition $topicPartition, " +
          s"but we only have log segments upto $endOffset.")

      if (startOffset == maxOffsetMetadata.messageOffset) {
        emptyFetchDataInfo(maxOffsetMetadata, includeAbortedTxns)
      } else if (startOffset > maxOffsetMetadata.messageOffset) {
       // Updated code to avoid the loop:
        val tmpFetchDataInfo = readFetchDataInfo(segmentOpt.get, startOffset, maxLength = 1, minOneMessage = false, nextOffsetMetadata, includeAbortedTxns = false)
        emptyFetchDataInfo(tmpFetchDataInfo.fetchOffsetMetadata, includeAbortedTxns)
      } else {
        readFetchDataInfo(segmentOpt.get, startOffset, maxLength, minOneMessage, maxOffsetMetadata, includeAbortedTxns)
      }
    }
  }

  private def readFetchDataInfo(segment: LogSegment,
                                startOffset: Long,
                                maxLength: Int,
                                minOneMessage: Boolean,
                                maxOffsetMetadata: LogOffsetMetadata,
                                includeAbortedTxns: Boolean): FetchDataInfo = {
    // Do the read on the segment with a base offset less than the target offset
    // but if that segment doesn't contain any messages with an offset greater than that
    // continue to read from successive segments until we get some messages or we reach the end of the log
    var fetchDataInfo: FetchDataInfo = null
    var segmentOpt: Optional[LogSegment] = Optional.of(segment)
    while (fetchDataInfo == null && segmentOpt.isPresent) {
      val segment = segmentOpt.get
      val baseOffset = segment.baseOffset

      val maxPosition =
        // Use the max offset position if it is on this segment; otherwise, the segment size is the limit.
        if (maxOffsetMetadata.segmentBaseOffset == segment.baseOffset) maxOffsetMetadata.relativePositionInSegment
        else segment.size

      fetchDataInfo = segment.read(startOffset, maxLength, maxPosition, minOneMessage)
      if (fetchDataInfo != null) {
        if (includeAbortedTxns)
          fetchDataInfo = addAbortedTransactions(startOffset, segment, fetchDataInfo)
      } else segmentOpt = segments.higherSegment(baseOffset)
    }

    if (fetchDataInfo != null) fetchDataInfo
    else {
      // okay we are beyond the end of the last segment with no data fetched although the start offset is in range,
      // this can happen when all messages with offset larger than start offsets have been deleted.
      // In this case, we will return the empty set with log end offset metadata
      new FetchDataInfo(nextOffsetMetadata, MemoryRecords.EMPTY)
    }
  }

```

Reply:
I don't think we need to change it because it won't cause infinite recursion in `LocalLog.read()`:
```
else if (startOffset > maxOffsetMetadata.messageOffset)
        emptyFetchDataInfo(convertToOffsetMetadataOrThrow(startOffset), includeAbortedTxns)
```

The reason is when in `LocalLog#convertToOffsetMetadataOrThrow`, the `maxOffsetMetadata` will be set to `nextOffsetMetadata`, which is the `logEndOffset`.

```
read(offset,
      maxLength = 1,
      minOneMessage = false,
      maxOffsetMetadata = nextOffsetMetadata,
      includeAbortedTxns = false)
```
So, even it enters the `LocalLog.read()` again, but this time, it'll enter different if/else block. It won't enter the `else if (startOffset > maxOffsetMetadata.messageOffset)` block this time because the check earlier will catch it!
```
      if (startOffset > endOffset || !segmentOpt.isPresent)
        throw new OffsetOutOfRangeException(s"Received request for offset $startOffset for partition $topicPartition, " +  s"but we only have log segments upto $endOffset.")
```

Reply:
> However, if nextOffsetMetadata doesn't change and somehow startOffset > maxOffsetMetadata.messageOffset, then we could loop forever.

I still don't know why the infinite loop will happen. In `Locallog#read`, we'll check if `startOffset > nextOffsetMetadata.messageOffset`, if so, `OffsetOutOfRangeException` will be thrown. 

```
  def read(startOffset: Long,
           maxLength: Int,
           minOneMessage: Boolean,
           maxOffsetMetadata: LogOffsetMetadata,
           includeAbortedTxns: Boolean): FetchDataInfo = {
....

      // (1) We will check if startOffset > nextOffsetMetadata.messageOffset, so if the provided `maxOffsetMetadata == nextOffsetMetadata`, the `OffsetOutOfRangeException` should be thrown before entering another `convertToOffsetMetadataOrThrow` loop
      val endOffsetMetadata = nextOffsetMetadata
      val endOffset = endOffsetMetadata.messageOffset
      var segmentOpt = segments.floorSegment(startOffset)

      // return error on attempt to read beyond the log end offset
      if (startOffset > endOffset || !segmentOpt.isPresent)
        throw new OffsetOutOfRangeException(s"Received request for offset $startOffset for partition $topicPartition,  but we only have log segments upto $endOffset.")

      if (startOffset == maxOffsetMetadata.messageOffset)
        emptyFetchDataInfo(maxOffsetMetadata, includeAbortedTxns)
      
     // So for the first time, the provided `maxOffsetMetadata` might be highWaterMark or LastStableOffset, it could enter this `else if` condition because HWM/LSO should is always less than LEO. But the 2nd time won't enter here.
      else if (startOffset > maxOffsetMetadata.messageOffset)
        emptyFetchDataInfo(convertToOffsetMetadataOrThrow(startOffset), includeAbortedTxns)
```
So there should be no chance that we do  `Locallog#read` with these `maxOffsetMetadata = nextOffsetMetadata` provided entering another `convertToOffsetMetadataOrThrow` loop.

If this did happen, it means the `logEndOffset < highWaterMarkOffset` or `logEndOffset < lastStableOffseet`, which I don't think this would happen.

Does that make sense?

Reply:
Thanks for the explanation. Yes, it makes sense to me now.